Issue58: Notifications: Step 1 :Persist User Personalization: DB change

ghormadesoham opened the pull request:
# Changes
- Add knex migration script. 
- Add schema details created using objection. 
- Key off schema shared by Mehmet. 
- Notably, use 1: many relationship for user to user personalization as done for user to review i.e one user can have many reviews as seen in Review.ts. Same way, one user can have many user personalizations.

# Testing
- verify that the website runs successfully using npm run dev and pages load. Verify that the DB has omscentral_user_personalization table
- 
# Related
- #2 graphql change 
- #4 server side change

ghormadesoham committed:
Issue58: DB change: Add knex migration script. Add schema detaills created using objection. Key off schema shared by Mehmet. Notably, use 1: many relationship for user to user personalization as done for user to review i.e one user can have many reviews as seen in Review.ts. Same way, one user can have many user personalizations.
Testing: verify that the website runs successfully using npm run dev and pages load. Verify that the DB has omscentral_user_personalization table

mehmetbajin commented on server/database/migrations/20211222089010_create_table_user_personalization.ts:
```diff
@@ -0,0 +1,23 @@
+      .inTable(User.tableName)
+      .index(`index_${UserPersonalization.tableName}_user_id`)
+      .onDelete('CASCADE');
+    tb.text('type').nullable();
```
Should `type` be `notNullable`?

When would it be set to NULL?

ghormadesoham replied:
My original understanding was that all non-primary key and non-foreign key constraints should be nullable.
For example, for a user, email is optional and hence [nullable](https://github.com/OMSCentral/omscentral/blob/69338c9ccf1fb6fae5b4ba5df57ce8de6d190822/server/database/migrations/20191116001050_create_table_user.ts#L14).

However, for user personalization, which is personalization data for a user, the type and value will not be nullable because other than type and value there is nothing else to store in terms of personalization for a given user.
Good point. 
As a next step, I will make both type and value to be nullable.
Hope the understanding that the same argument applies to value makes sense.
Thanks!

ghormadesoham replied:
Done!

mehmetbajin commented on server/src/models/UserPersonalization.ts:
```diff
@@ -0,0 +1,43 @@
+
+  static relationMappings = {
+    user: {
+      relation: Domain.HasOneRelation,
```
Prefer `Domain.BelongsToOneRelation` instead, since a given personalization belongs to a single user.

ghormadesoham replied:
Per your comment above, my understanding is that Domain.HasOneRelation and Domain.BelongsToOneRelation are equivalent.
However, per example in objection [documentation](https://vincit.github.io/objection.js/guide/relations.html#examples) they are different.
Quoting:
"BelongsToOneRelation: Use this relation when the source model has the foreign key
HasManyRelation: Use this relation when the related model has the foreign key."
However, importantly, since both are one-one relationship, I will switch to Domain.BelongsToOneRelation in this case.

mehmetbajin reviewed (changes requested):
A couple questions.

ghormadesoham committed:
DB:Code review:
1.columns type and value should not be nullable since they will always be specified whenever a new row is added to UserPersonalization table.
2. Use BelongsToOneRelation instead of HasOneRelation since a user personalization belongs to a user. i.e both are 1:1 but BelongsToOneRelation makes sense in this context.

ghormadesoham commented on server/src/models/UserPersonalization.ts:
```diff
@@ -0,0 +1,43 @@
+    properties: {
+      id: { type: 'string' },
+      user_id: { type: 'string' },
+      type: { type: ['string', 'null'] },
```
Remove null from jsonSchema.

ghormadesoham commented on server/src/models/UserPersonalization.ts:
```diff
@@ -0,0 +1,43 @@
+    },
+  };
+
+  static eagerQuery = (): QueryBuilder<UserPersonalization> =>
```
The eager query should be right since it was keyed off Course to CourseMetric which is also 1: many.

Source [code](https://github.com/OMSCentral/omscentral/blob/69338c9ccf1fb6fae5b4ba5df57ce8de6d190822/server/src/models/Course.ts#L56) for reference:
```
  static relationMappings = {
    metric: {
      relation: Domain.HasOneRelation, // HasOneRelation is preferred over BelongsToOneRelation since
// a course has a course metric instead of metric belongs to a course
      modelClass: CourseMetric,
      join: {
        from: `${Course.tableName}.id`,
        to: `${CourseMetric.tableName}.course_id`,
      },
    },
  };
...
  static eagerQuery = (): QueryBuilder<Course> =>
    Course.query().withGraphFetched('[metric]');
```

ghormadesoham committed:
DB: Make type and value as required since they are not nullable.
Reference: See Specialization table in server/src/models and its migration file in server/database/migrations

ghormadesoham commented on server/src/models/UserPersonalization.ts:
```diff
@@ -0,0 +1,43 @@
+  static tableName = 'omscentral_user_personalization';
+
+  static relationMappings = {
+    user: {
```
```user``` should be ```userPersonalization``` to match the table defined in  the ```from``` field.
This is necessary to resolve the attached error:
[BelongsToOneRelation stack trace Dec 26.txt](https://github.com/ghormadesoham/omscentral/files/7780233/BelongsToOneRelation.stack.trace.Dec.26.txt)

ghormadesoham replied:
Done!

ghormadesoham commented on server/src/models/UserPersonalization.ts:
```diff
@@ -0,0 +1,43 @@
+import { User } from './User';
+import { withDates } from './utils';
+
+export class UserPersonalization extends withDates(Domain) {
```
```
export class UserPersonalization extends withDates(Domain)
```
should be 
```
export class UserPersonalization extends Domain
```
since UserPersonalization does not require created column which User required.

ghormadesoham replied:
Done!

ghormadesoham committed:
Issue58: DB:1.Inherit UserPersonalization from Domain directly. This is important because you do not need the created column which becomes required once you inherit from WithDates.
2.Fix upsert operation error due to the relationMapping not specifying source table to be same as from field in join argument.
